### PR TITLE
Add support for --env-host

### PR DESCRIPTION
**Pull request by rhatdan**
This flag passes the host environment into the container.  The basic idea is to
leak all environment variables from the host into the container.

Environment variables from the image, and passed in via --env and --env-file
will override the host environment.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@adrianreber PTAL
@baude @mheon @vrothberg @giuseppe @ashley-cui @QiWang19 @TomSweeneyRedHat PTAL

File: cmd/podman/common.go [hunk 1]
```diff
@@ -221,6 +221,9 @@ func getCreateFlags(c *cliconfig.PodmanCommand) {
 		"env", "e", []string{},
 		"Set environment variables in container",
 	)
+	createFlags.Bool(
+		"env-host", false, "Use all current host environment variables in container",
```
Comment (mheon) [inline cmd/podman/common.go]: Should this be `host-env` instead?

File: cmd/podman/shared/create.go [hunk 2]
```diff
@@ -483,6 +483,16 @@ func ParseCreateOpts(ctx context.Context, c *GenericCLIResults, runtime *libpod.
 
 	// ENVIRONMENT VARIABLES
 	env := EnvVariablesFromData(data)
+	if c.Bool("env-host") {
+		for _, e := range os.Environ() {
+			pair := strings.SplitN(e, "=", 2)
+			if _, ok := env[pair[0]]; !ok {
```
Comment (mheon) [inline cmd/podman/shared/create.go]: Shouldn't we overwrite if the goal is to send over the entire host environment?

Comment (rhatdan) [reply cmd/podman/shared/create.go]: No I believe the precidence should be from the host, image, env-file, env in that order.  That is what I documented in the man page.

File: cmd/podman/common.go (hunk 1 above)
Comment (rhatdan) [reply cmd/podman/common.go]: I wanted to  keep all of the --env* flags the same, so they would be easier to discover.

**Comment by rhatdan** [comment]
@adrianreber Discovered a use case for this flag when running in MPI

**Comment by AkihiroSuda** [comment]
Wondering this should support exposing only variables with specific prefix string (eg. MPI_), for security reason

**Comment by adrianreber** [comment]
> Wondering this should support exposing only variables with specific prefix string (eg. MPI_), for security reason

For what I am trying to do I only need variables with PMIX_ and OMPI_. If there is an option to specify the prefix it would also work. Having something which just passes everything into Podman would be the easiest. But I am happy with both ways (all variables or only some).

I will try this PR to see if it works for my mpirun use case.

**Comment by adrianreber** [comment]
Besides the broken test case this is exactly what I was looking for. Thanks.

I tried it and now I can run an Open MPI program (rootless) on a system without InfiniBand:
```
$ mpirun --hostfile ~/hosts --mca orte_tmpdir_base /tmp/podman-mpirun podman run --env-host --security-opt label=disable -v /tmp/podman-mpirun:/tmp/podman-mpirun --userns=keep-id --net=host mpi-test /home/mpi/ring
Rank 0 has cleared MPI_Init
Rank 1 has cleared MPI_Init
Rank 0 has completed ring
Rank 0 has completed MPI_Barrier
Rank 1 has completed ring
Rank 1 has completed MPI_Barrier
```

I am using ring.c from Open MPI: https://raw.githubusercontent.com/open-mpi/ompi/master/orte/test/mpi/ring.c

This also works for InfiniBand based system by including the InfiniBand devices into the container:
```
$ mpirun --mca btl ^openib --hostfile ~/hosts --mca orte_tmpdir_base /tmp/podman-mpirun podman run --env-host -v /tmp/podman-mpirun:/tmp/podman-mpirun --security-opt label=disable --userns=keep-id --device /dev/infiniband/uverbs0 --device /dev/infiniband/umad0 --device /dev/infiniband/rdma_cm --net=host mpi-test /home/mpi/ring
Rank 0 has cleared MPI_Init
Rank 1 has cleared MPI_Init
Rank 0 has completed ring
Rank 0 has completed MPI_Barrier
Rank 1 has completed ring
Rank 1 has completed MPI_Barrier
```

I am running my tests with OpenMPI 4.0.1 on RHEL 8.

**Comment by rhatdan** [comment]
A syntax of --env PMIX_* and --env OMPI_*, would be something I would consider.  This would give you a little more fine grain control, and would match up closely to what we already do,  but I still think adding --env-host is worth doing.

**Comment by rhatdan** [comment]
@mheon @vrothberg @giuseppe @ashley-cui @QiWang19 @TomSweeneyRedHat @baude PTAL

File: docs/podman-create.1.md [hunk 3]
```diff
@@ -901,6 +903,19 @@ The fuse-overlay package provides a userspace overlay storage driver, otherwise
 the vfs storage driver, which is diskspace expensive and does not perform well. slirp4netns is
 required for VPN, without it containers need to be run with the --net=host flag.
 
+## ENVIRONMENT
+
+Environment variables within containers can be set using multiple different options:  This section describes the presidence.
+
+Presidence Order:
```
Comment (rst0git) [inline docs/podman-create.1.md]: s/Presidence/Precedence/

File: docs/podman-create.1.md [hunk 4]
```diff
@@ -901,6 +903,19 @@ The fuse-overlay package provides a userspace overlay storage driver, otherwise
...
+	   **--env-host** : Host environment of the process executing podman is added.
+
+	   Container image : Any enviroment variables specified in the contianer image.
```
Comment (rst0git) [inline docs/podman-create.1.md]: s/contianer/container/

File: docs/podman-create.1.md [hunk 5]
```diff
@@ -901,6 +903,19 @@ The fuse-overlay package provides a userspace overlay storage driver, otherwise
...
+
+	   **--env-file** : Any environment variables specfied via env-files.  If multiple files specified, then they override each other in order of entry.
```
Comment (rst0git) [inline docs/podman-create.1.md]: s/specfied/specified/

File: docs/podman-create.1.md [hunk 6]
```diff
@@ -901,6 +903,19 @@ The fuse-overlay package provides a userspace overlay storage driver, otherwise
...
+
+	   **--env** : Any environment variables specified will overide previous settings.
```
Comment (rst0git) [inline docs/podman-create.1.md]: s/overide/override/

Comment (rhatdan) [reply docs/podman-create.1.md]: fixed

File: docs/podman-create.1.md (hunk 5 above)
Comment (rhatdan) [reply docs/podman-create.1.md]: Fixed

File: docs/podman-create.1.md (hunk 3 above)
Comment (rhatdan) [reply docs/podman-create.1.md]: Fixed

**Comment by rhatdan** [comment]
Added  a new patch to go through and fix all spelling mistakes in documentation.

**Comment by rhatdan** [comment]
@mheon @vrothberg @giuseppe @ashley-cui @QiWang19 @TomSweeneyRedHat @baude PTAL

**Comment by giuseppe** [comment]
LGTM

**Comment by adrianreber** [comment]
Thanks for the quick code changes. I mentioned this in my mail thread on the Open MPI user mailing list: https://www.mail-archive.com/users@lists.open-mpi.org/msg33342.html

For completeness I also want to mention here that intra node communication based on shared memory still does not work with Podman and Open MPI. So currently I have to disable the shared memory transport layer in Open MPI by telling `mpirun` to not use shared memory communication: `--mca btl ^vader`. I am guessing that Open MPI then switches to TCP for localhost communication.

**Comment by rhatdan** [comment]
@adrianreber There is no reason to volume mount in /dev/shm with --ipc=host.  --ipc=host does this automatically.

```
# touch /dev/shm/dan
# podman run --ipc=host fedora ls -l /dev/shm/dan
-rw-r--r--. 1 root root 0 Jul 13 10:55 /dev/shm/dan
```

**Pull request merged**